nixos/modules/virtualisation/containers-next/default.nix

```diff
@@ -0,0 +1,571 @@
+              ++ optionals (config.network.v4.static.hostAddresses != null)
+                config.network.v4.static.hostAddresses
+              ++ optionals (config.network.v6.static.hostAddresses != null)
+                config.network.v6.static.hostAddresses;
+            networkConfig = mkNetworkCfg (config.network.v4.addrPool != []) config.network.v4.nat;
+          };
+        }))
+        ++ (flip mapAttrsToList config.nixos.containers.zones (name: zone: {
+          "20-${ifacePrefix "zone"}-${name}" = {
+            matchConfig = mkMatchCfg "zone" name;
+            address = zone.v4.addrPool
+              ++ zone.v6.addrPool
+              ++ zone.hostAddresses
+              ++ (flatten (flip mapAttrsToList cfg
+                (name: config: optionals (config.zone != null && config.network != null)
+                  (config.network.v4.static.hostAddresses ++ config.network.v6.static.hostAddresses)
+                )));
+            networkConfig = mkNetworkCfg true zone.v4.nat;
+          };
+        })));
+
+      nspawn = mapAttrs (const mkContainer) images;
+      targets.machines.wants = map (x: "systemd-nspawn@${x}.service") (attrNames cfg);
+      services = listToAttrs (flip map (attrNames cfg) (container:
+        let
+          inherit (cfg.${container}) activation credentials;
+        in nameValuePair "systemd-nspawn@${container}" {
+          preStart = mkBefore ''
+            if [ ! -d /var/lib/machines/${container} ]; then
+              mkdir -p /var/lib/machines/${container}/{etc,var,nix/var/nix}
+              touch /var/lib/machines/${container}/etc/{os-release,machine-id} || true
+            fi
+          '';
+
+          partOf = [ "machines.target" ];
+          before = [ "machines.target" ];
+
+          serviceConfig = mkMerge [
+            {
+              # Inherit settings from `systemd-nspawn@.service`.
+              # Workaround since settings from `systemd-nspawn@.service`-settings are not
+              # picked up if an override exists and `systemd-nspawn@ldap` exists.
+              RestartForceExitStatus = 133;
+              Type = "notify";
+              TasksMax = 16384;
+              WatchdogSec = "3min";
+              SuccessExitStatus = 133;
+              Delegate = "yes";
+              KillMode = "mixed";
+              Slice = "machine.slice";
+              DevicePolicy = "closed";
+              DeviceAllow = [
+                "/dev/net/tun rwm"
+                "char-pts rw"
+                "/dev/loop-control rw"
+                "block-loop rw"
+                "block-blkext rw"
+                "/dev/mapper/control rw"
+                "block-device-mapper rw"
+              ];
+              X-ActivationStrategy = activation.strategy;
+              ExecStart = [
+                ""
+                "${config.systemd.package}/bin/systemd-nspawn ${credentials} --quiet --keep-unit --boot --network-veth --settings=override --machine=%i"
+              ];
```

Comment:
I think we should make TimeoutStartSec configurable. Some services like hound-search can take a bit to start up and if the container does not reach multi-user.target in the default timeout of I think 30 seconds it gets completely killed.

Also I think we want to set ``wantedBy=multi-user.target`` to automatically start it on system reboot.

Reply:
> Also I think we want to set wantedBy=multi-user.target to automatically start it on system reboot.

This is already done by adding each container to `machines.target` which seems to be added by `systemd` for exactly that purpose, so it seemed to be the more intuitive choice.

> I think we should make TimeoutStartSec configurable

IIRC I just copied over upstream defaults. But I'll check that.

Reply:
Hmm, may I ask what's wrong with setting `systemd.services."systemd-nspawn@foo".serviceConfig.TimeoutStartSec` as needed?

Reply:
> Hmm, may I ask what's wrong with setting `systemd.services."systemd-nspawn@foo".serviceConfig.TimeoutStartSec` as needed?

It took me a longer time to figure that out and I think it would be nice if it would be easier to set.

Reply:
The current container implementation has this option as well: `containers.<name>.timeoutStartSec`. I believe it was @davidak who felt the same and had this option implemented, IIRC.

Reply:
Yes, it was needed to be able to start a lot of containers. https://github.com/NixOS/nixpkgs/pull/65661

Reply:
If we want to keep it simple maybe just document this at a good place?